From 65869d827a7832e9874782523744066f10aa9834 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Wed, 22 May 2024 15:20:49 +0200
Subject: [PATCH 0945/1695] drivers: stm32mp_provisioning: fix unpaged
 low-power sequence

Change implementation to ensure there is no access to the DTB during
low power sequence. MP15 cannot support that because it would mean the
embedded DTB is saved in pager resident memory.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git ff2093b3e4b2a7048ed4ee6e11fa7d23b25714ca drivers: stm32mp_provisioning: fix unpaged low-power sequence]
Change-Id: I946723c656960d68d59493d0c1bba87d6f5c12e7
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/382371
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Tested-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/stm32mp_provisioning.c | 190 ++++++++++++++++++----------
 1 file changed, 124 insertions(+), 66 deletions(-)

diff --git a/core/drivers/stm32mp_provisioning.c b/core/drivers/stm32mp_provisioning.c
index 9011e4fee..8528d8f8c 100644
--- a/core/drivers/stm32mp_provisioning.c
+++ b/core/drivers/stm32mp_provisioning.c
@@ -17,15 +17,103 @@
 #include <stdbool.h>
 #include <stdlib.h>
 #include <stm32_util.h>
+#include <sys/queue.h>
 #include <trace.h>
 
-static TEE_Result provisioning_subnode(const void *fdt, int node)
+/*
+ * struct shadow_otp - OTP shadow memory cells provisioning
+ * @otp_id: OTP base offset identifier
+ * @otp_lock: OTP locking mask
+ * @otp_len_word: Number of cells of @value
+ * @value: Arrays of the BSEC words to load in shadow memory
+ * @link: Reference in list
+ */
+struct shadow_otp {
+	uint32_t otp_id;
+	uint32_t otp_lock;
+	size_t otp_len_word;
+	uint32_t *value;
+	SLIST_ENTRY(shadow_otp) link;
+};
+
+SLIST_HEAD(shadow_otp_head, shadow_otp);
+static struct shadow_otp_head shadow_otp_head =
+	SLIST_HEAD_INITIALIZER(&shadow_otp_head);
+
+static void provision(struct shadow_otp *shadow_otp, size_t index)
 {
+	uint32_t provis_value = shadow_otp->value[index];
+	uint32_t provis_lock = shadow_otp->otp_lock;
+	uint32_t otp_id = shadow_otp->otp_id + index;
+	uint32_t otp_val = 0;
+	bool otp_lock = false;
+
+	if (stm32_bsec_read_sw_lock(otp_id, &otp_lock))
+		panic();
+	if (stm32_bsec_shadow_read_otp(&otp_val, otp_id))
+		panic();
+
+	DMSG("OTP %"PRIu32": provisioning value %#"PRIx32", lock %#"PRIx32
+	     " / shadow value %#"PRIx32", lock %#"PRIx32,
+	     otp_id, provis_value, provis_lock, otp_val, otp_lock);
+
+	if (otp_val != (otp_val | provis_value)) {
+		if (otp_lock) {
+			EMSG("Override the OTP %"PRIu32": shadow write lock",
+			     otp_id);
+			panic();
+		}
+
+		IMSG("Override the OTP %"PRIu32": %#"PRIx32" to %#"PRIx32,
+		     otp_id, otp_val, otp_val | provis_value);
+
+		if (stm32_bsec_write_otp(otp_val | provis_value, otp_id))
+			panic();
+	}
+
+	switch (provis_lock) {
+	case STICKY_LOCK_SW:
+		if (stm32_bsec_set_sw_lock(otp_id))
+			panic();
+		break;
+	case STICKY_LOCK_SR:
+		if (stm32_bsec_set_sr_lock(otp_id))
+			panic();
+		break;
+	case STICKY_LOCK_SWSR:
+		if (stm32_bsec_set_sw_lock(otp_id) ||
+		    stm32_bsec_set_sr_lock(otp_id))
+			panic();
+		break;
+	case STICKY_NO_LOCK:
+	default:
+		break;
+	}
+
+	if (stm32_bsec_shadow_read_otp(&otp_val, otp_id))
+		panic();
+	DMSG("Read SHADOW %#"PRIx32, otp_val);
+}
+
+static void load_provisioning(void)
+{
+	struct shadow_otp *shadow_otp = NULL;
+	size_t i = 0;
+
+	SLIST_FOREACH(shadow_otp, &shadow_otp_head, link) {
+		for (i = 0; i < shadow_otp->otp_len_word; i++)
+			provision(shadow_otp, i);
+	}
+}
+
+static void parse_dt(const void *fdt, int node)
+{
+	struct shadow_otp *shadow_otp = NULL;
 	TEE_Result res = TEE_ERROR_GENERIC;
 	int child = -1;
 
 	fdt_for_each_subnode(child, fdt, node) {
-		const int *cuint = NULL;
+		const uint32_t *cuint = NULL;
 		int len = 0;
 		uint32_t phandle = 0;
 		uint32_t otp_lock = 0;
@@ -44,6 +132,10 @@ static TEE_Result provisioning_subnode(const void *fdt, int node)
 		if (res)
 			panic("Phandle not found");
 
+		shadow_otp = calloc(1, sizeof(*shadow_otp));
+		if (!shadow_otp)
+			panic();
+
 		cuint = fdt_getprop(fdt, child, "st,shadow-lock", &len);
 		if (cuint && len > 0)
 			otp_lock = fdt32_to_cpu(*cuint);
@@ -56,65 +148,31 @@ static TEE_Result provisioning_subnode(const void *fdt, int node)
 		if (otp_bit_len % (sizeof(uint32_t) * CHAR_BIT))
 			otp_len_word++;
 
-		if ((unsigned int)len / sizeof(uint32_t) != otp_len_word)
-			panic("Invalid OTP size");
-
-		for (i = 0; i < otp_len_word; i++) {
-			uint32_t shadow_val = 0;
-			uint32_t otp_val = 0;
-			bool lock = false;
-
-			if (stm32_bsec_read_sw_lock(otp_id + i, &lock))
-				panic();
-
-			if (lock) {
-				EMSG("Override the OTP %u: shadow write lock",
-				     otp_id + i);
-				panic();
-			}
-
-			shadow_val = fdt32_to_cpu(*cuint++);
-
-			if (stm32_bsec_shadow_read_otp(&otp_val, otp_id + i))
-				panic();
-
-			if (otp_val != (otp_val | shadow_val)) {
-				IMSG("Override the OTP %u: %x to %x",
-				     otp_id + i, otp_val, otp_val | shadow_val);
-
-				if (stm32_bsec_write_otp(otp_val | shadow_val,
-							 otp_id + i))
-					panic();
-			}
-
-			switch (otp_lock) {
-			case STICKY_LOCK_SW:
-				if (stm32_bsec_set_sw_lock(otp_id + i))
-					panic();
-				break;
-			case STICKY_LOCK_SR:
-				if (stm32_bsec_set_sr_lock(otp_id + i))
-					panic();
-				break;
-			case STICKY_LOCK_SWSR:
-				if (stm32_bsec_set_sw_lock(otp_id + i) ||
-				    stm32_bsec_set_sr_lock(otp_id + i))
-					panic();
-				break;
-			case STICKY_NO_LOCK:
-			default:
-				break;
-			}
-
-			stm32_bsec_shadow_read_otp(&otp_val, otp_id + i);
-			DMSG("Read SHADOW %x", otp_val);
+		if ((unsigned int)len / sizeof(uint32_t) != otp_len_word) {
+			EMSG("Invalid size %d bytes for OTP %"PRIu32" (%zu words)",
+			     len, otp_id, otp_len_word);
+			panic();
 		}
-	}
 
-	return res;
+		if (!otp_len_word)
+			continue;
+
+		shadow_otp->value = calloc(otp_len_word, BSEC_BYTES_PER_WORD);
+		if (!shadow_otp->value)
+			panic();
+
+		shadow_otp->otp_id = otp_id;
+		shadow_otp->otp_lock = otp_lock;
+		shadow_otp->otp_len_word = otp_len_word;
+
+		for (i = 0; i < otp_len_word; i++)
+			shadow_otp->value[i] = fdt32_to_cpu(*cuint++);
+
+		SLIST_INSERT_HEAD(&shadow_otp_head, shadow_otp, link);
+	}
 }
 
-static TEE_Result provisioning_init(void)
+static void provisioning_init(void)
 {
 	const void *fdt = get_embedded_dt();
 	const char __maybe_unused *name = NULL;
@@ -125,36 +183,36 @@ static TEE_Result provisioning_init(void)
 
 	node = fdt_node_offset_by_compatible(fdt, 0, "st,provisioning");
 	if (node < 0)
-		return TEE_SUCCESS;
+		return;
 
 	if (fdt_first_subnode(fdt, node) == -FDT_ERR_NOTFOUND) {
 		name = fdt_get_name(fdt, node, NULL);
 		DMSG("Warning : no subnode in %s\n", name);
-		return TEE_SUCCESS;
+		return;
 	}
 
-	return provisioning_subnode(fdt, node);
+	parse_dt(fdt, node);
+	load_provisioning();
 }
 
 static TEE_Result provisioning_pm(enum pm_op op, unsigned int pm_hint __unused,
 				  const struct pm_callback_handle *hdl __unused)
 {
 	if (op == PM_OP_RESUME)
-		return provisioning_init();
+		load_provisioning();
 
 	return TEE_SUCCESS;
 }
+DECLARE_KEEP_PAGER(provisioning_pm);
 
 static TEE_Result provisioning_probe(void)
 {
-	TEE_Result res = TEE_ERROR_GENERIC;
+	provisioning_init();
 
-	res = provisioning_init();
-
-	if (res == TEE_SUCCESS && IS_ENABLED(CFG_PM))
+	if (!SLIST_EMPTY(&shadow_otp_head))
 		register_pm_driver_cb(provisioning_pm, NULL, "provisioning");
 
-	return res;
+	return TEE_SUCCESS;
 }
 
 early_init_late(provisioning_probe);
-- 
2.47.2

